whylog/teacher/__init__.py

```diff
@@ -45,20 +105,57 @@ def guess_pattern(self, line_id):
     def set_pattern_name(self, line_id, name):
         pass
 
-    def set_converter(self, line_id, group, converter):
+    def set_converter(self, pattern_group, converter):
         pass
 
-    def set_primary_key(self, line_id, groups):
+    def set_primary_key(self, line_id, group_numbers):
         pass
 
     def set_log_type(self, line_id, log_type):
         pass
 
-    def register_constraint(self, constr_id, constr_type, groups, params=None):
-        pass
+    def register_constraint(self, constr_id, pattern_groups, constraint):
+        """
+        Adds new constraint to rule.
 
-    def remove_constraint(self, constr_id):
-        pass
+        If constr_id already exists, constraint with this constr_id
+        is overwritten by new constraint
+
+        :param pattern_groups: groups in pattern which contain text params
+                               that are linked by constraint
+        :type pattern_groups: PatternGroup list
+        """
+        if constr_id in self._constraint_base.keys():
+            self.remove_constraint(constr_id)
+
+        self._constraint_base[constr_id] = constraint
+        new_constraint_links = [
+            (group.line_id, group.number, constr_id) for group in pattern_groups
+        ]
+        self._constraint_links.insert_links(new_constraint_links)
+
+    def remove_constraint(self, removing_constr_id):
+        """
+        Removes constraint from rule.
+
+        Assumption: Constraint already exists in rule.
+        """
+        self._constraint_links.remove_links_where(constr_id=removing_constr_id)
+        del self._constraint_base[removing_constr_id]
+
+    def _remove_constraints_by_line(self, line_id):
+        self._constraint_links.remove_links_where(line_id=line_id)
+        self._sync_constraint_base_with_links()
+
+    def _remove_constraint_by_group(self, group):
+        self._constraint_links.remove_links_where(line_id=group.line_id, group_no=group.number)
+        self._sync_constraint_base_with_links()
+
+    def _sync_constraint_base_with_links(self):
+        ids_from_links = self._constraint_links.distinct_constraint_ids()
```

Comment:
In my opinion it more like teacher function, because it modifies Teacher's field - self._constraint_base